design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,186 @@
+host status by re-inspecting the host, but it will force the already provisioned
+host to go through another round of provisioning, which eventually reboots the
+host after IPA steps are completed. To have the inspection data available in the
+target cluster and to avoid unnecessary provisioning after pivoting, we currently
+copy inspection data to baremetalhost.metal3.io/status annotation so that BMO in
+the target cluster can continue reconciliation from a point it was before pivoting,
+avoiding redundant provisioning.
+
+However, there is a limitation with the current approach, because when inspection
+data is quite big (e.g. when the host has 1000 multipath disks), it is easy to hit
+the global limit of `metadata.annotations` [262144](https://github.com/kubernetes/kubernetes/blob/master/test/integration/controlplane/synthetic_controlplane_test.go#L313) bytes and fail to pivot successfully.
+
+### Goals
+
+- Store host inspection data into InspectionData CR.
+- Rely on InspectionData instead of status annotation when pivoting.
+
+### Non-Goals
+
+Remove `status.hardware` from BareMetalHost API
+
+## Proposal
+
+### User Stories
+
+As an operator, from an ephemeral cluster I would like to pivot/move to a target
+cluster successfully.
+
+## Design Details
+
+Apart of writing hardware details into the status of BareMetalHost, write it
+to a InspectionData CR spec. Baremetalhost and InspectionData will be parent
+and child objects respectively, and this relationship is controlled via
+ownerReferences chain. As such, the same name & namespace will be used for
+InspectionData as its owning object BareMetalHost. The spec of InspectionData
+will contain the same hardware information as currently under `status.hardware`
+of BareMetalHost.
+
+### Pivoting
+
+There are a couple of options one can choose to pivot objects to a target cluster
+that are outside of the Cluster API chain.
+
+1. [`clusterctl.cluster.x-k8s.io/move-hierarchy`](https://github.com/kubernetes-sigs/cluster-api/blob/0e6428151f2a388b98d1e220a698b85940a5f290/cmd/clusterctl/api/v1alpha3/labels.go#L38) label should be set
+    on BareMetalHost CRD before issuing *`clusterctl move`* so that the entire hierarchy
+    (i.e. including InspectionData, Secrets and ConfigMaps) is pivoted to the target
+    cluster.
+
+2. `clusterctl.cluster.x-k8s.io=""` and `cluster.x-k8s.io/provider="metal3"`
+    labels on InspectionData CRD to extend the list of provider objects.
+
+3. `clusterctl.cluster.x-k8s.io/move` label on InspectionData CRD. This will
+    not move child objects as in option 1.
+
+### Implementation Details/Notes/Constraints
+
+The life of an InspectionData is tightly coupled to a BareMetalHost. Creating or
+deleting a BareMetalHost will result in creation or deletion of a InspectionData
+respectively. Deprovisioning of the BareMetalHost will result in the inspection
+data removal from the spec of InspectionData.
```

Comment:
> I don't think we should bother to create the CR until we do inspection. 

When we create a BareMetalHost, it automatically goes through inspection state until it reaches available right? If yes, why should not we create the CR whenever BareMetalHost is created? 

> We definitely must not remove the inspection data on deprovisioning. 

Sorry my bad, I had a wrong impression that we delete the inspection data when deprovisioning. Though, related to immutability point above, if the CR is immutable, and if someone requests inspection (due to some hardware change) of `Available` BMH, we can't update the CR. So, then we delete and re-create it ?
 
> We should add a finalizer on the CR to prevent it being deleted while a Host is provisioned. More details on how I think the lifecycle should work: https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.vb3zlo2h9t7

+1

Reply:
> > I don't think we should bother to create the CR until we do inspection.
> 
> When we create a BareMetalHost, it automatically goes through inspection state until it reaches available right? If yes, why should not we create the CR whenever BareMetalHost is created?

Eventually, but there are a couple of states to go through first. If we get stuck in e.g. Registration there is probably no reason to make it look like we've already done inspection by creating the CR. Also, there's an annotation you can add to skip inspection.

> > We definitely must not remove the inspection data on deprovisioning.
> 
> Sorry my bad, I had a wrong impression that we delete the inspection data when deprovisioning. Though, related to immutability point above, if the CR is immutable, and if someone requests inspection (due to some hardware change) of `Available` BMH, we can't update the CR. So, then we delete and re-create it ?

Yes. I think this will make implementation easier too.

> > We should add a finalizer on the CR to prevent it being deleted while a Host is provisioned. More details on how I think the lifecycle should work: https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.vb3zlo2h9t7
> 
> +1